File: Week3/Code/Vectorize2.py
#! usr/bin/python

""" Author - Oliver Tarrant
A python version of Vectorize2.R. A vectorized version of the 
stochastic Ricker model"""

import scipy as sc
import time
import timeit


def stochrick(p0=sc.random.uniform(0.5,1.5,1000),r=1.2,k=1,sigma=0.2,numyears=100):
	N = sc.empty(shape=(numyears, len(p0)))
	N[0,]=p0
	for pop in range(len(p0)):
		for yr in range(1,numyears):
			N[yr,pop] = N[yr-1,pop]*sc.exp(r*(1-N[yr-1,pop]/k)+sc.random.normal(0,sigma,1))  #(mean,sd,n)
			
	return N 
start = time.time()
stochrick()
print "stockrick takes %f s to run!" %(time.time()-start)

def stochrickvect(p0=sc.random.uniform(0.5,1.5,1000),r=1.2,k=1,sigma=0.2,numyears=100):
	N = sc.empty(shape=(numyears,len(p0)))
	N[0,]=p0
	for yr in range(1,numyears):
		N[yr,] = N[yr-1,]*sc.exp(r*(1-N[yr-1,]/k)+sc.random.normal(0,sigma,1))
		
	return N
start = time.time()	
stochrickvect()
print "stochrickvect takes %f s to run!" %(time.time()-start)	

File: Week3/Code/Vectorize2.R
# Runs the stochastic (with gaussian fluctuations) Ricker Eqn .

rm(list=ls())

stochrick<-function(p0=runif(1000,.5,1.5),r=1.2,K=1,sigma=0.2,numyears=100)
{ # p0 is a vector of 1000 random numbers from uniform dist between 0.5 and 1.5
  #initialize
  N<-matrix(NA,numyears,length(p0)) # N is a matrix of NA with numyears rows and len(p0) columns
  N[1,]<-p0
  
  for (pop in 1:length(p0)) #loop through the populations 
  {
    for (yr in 2:numyears) #for each pop, loop through the years
    {
      N[yr,pop]<-N[yr-1,pop]*exp(r*(1-N[yr-1,pop]/K)+rnorm(1,0,sigma))
    } # Calculate the value of the next population step 
  }
  return(N)
}

 print("Stochastic Ricker takes:")
 print(system.time(res2<-stochrick()))
# Now write another code called stochrickvect that vectorizes the above 
# to the extent possible, with improved performance: 


rm(list=ls())

stochrickvect<-function(p0=runif(1000,.5,1.5),r=1.2,K=1,sigma=0.2,numyears=100)
{ # p0 is a vector of 1000 random numbers from uniform dist between 0.5 and 1.5
  #initialize
  N<-matrix(NA,numyears,length(p0)) # N is a matrix of NA with numyears rows and len(p0) columns
  N[1,]<-p0
   
  
    for (yr in 2:numyears) #for each pop, loop through the years
    {
      N[yr,]<-N[yr-1,]*exp(r*(1-N[yr-1,]/K)+rnorm(length(p0),0,sigma))
    } # Calculate the value of the next population step 
  
  return(N)

}
# Now write another code called stochrickvect that vectorizes the above 
# to the extent possible, with improved performance: 

 print("Vectorized Stochastic Ricker takes:")
 print(system.time(res2<-stochrickvect()))

 